### ISSUE TITLE
☂️-Issue for "Support for Kubernetes v1.26”

### ISSUE THREAD

**Issue by rfranzke** [gardener/gardener#6773]
/area open-source
/kind enhancement
/priority 3
/topology garden seed shoot

Kubernetes v1.26 was released on December 9th, you'll find the changelog here: https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG/CHANGELOG-1.26.md

Work can be started and pushed to the [`kubernetes-v1.26`](https://github.com/gardener/gardener/tree/kubernetes-v1.26) branch. There should be similar branches in the provider extension repositories in case the work has been started already.

---

# Tasks

## Hyperkube

- [x] `hyperkube`: https://github.com/gardener/hyperkube/releases/tag/v1.26.0
  - Build new image for v1.26

## Gardener

- [x] `gardener`: https://github.com/gardener/gardener/pull/7275
  - Good to know/no action required
    - Deprecated beta APIs scheduled for removal in `v1.26` are no longer served. See https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-26 for more information. (https://github.com/kubernetes/kubernetes/pull/111973)
    - `PodDisruptionBudget` now adds an alpha `spec.unhealthyPodEvictionPolicy` field. When the `PDBUnhealthyPodEvictionPolicy` feature-gate is enabled in `kube-apiserver`, setting this field to `"AlwaysAllow"` allows pods to be evicted if they do not have a ready condition, regardless of whether the PodDisruptionBudget is currently healthy.' (https://github.com/kubernetes/kubernetes/pull/113375)
    - A new auth API to get self subject attributes (new selfsubjectreviews API) is added. (https://github.com/kubernetes/kubernetes/pull/111333)
    - The is a new feature that allows a StatefulSet to start numbering replicas from an arbitrary non-negative ordinal, using the `.spec.ordinals.start` field. (https://github.com/kubernetes/kubernetes/pull/112744)
    - There is a new kube-proxy flag (`--iptables-localhost-nodeports`, default true) to allow disabling NodePort services on loopback addresses. Note: this only applies to iptables mode and ipv4. (https://github.com/kubernetes/kubernetes/pull/108250)
    - kube-controller-manager supports '--concurrent-horizontal-pod-autoscaler-syncs' flag to set the number of horizontal pod autoscaler controller workers. Previously the workers count was 1 as was not configurable. (https://github.com/kubernetes/kubernetes/pull/108501)
    - In 'kube-proxy`: The "userspace" proxy mode (deprecated for over a year) is no longer supported on either Linux or Windows. Users should use "iptables" or "ipvs" on Linux, or "kernelspace" on Windows. (https://github.com/kubernetes/kubernetes/pull/112133)
    - Legacy klog flags are no longer available. Only -v and -vmodule are still supported. (https://github.com/kubernetes/kubernetes/pull/112120)
    - The `DynamicKubeletConfig` feature gate has been removed from the API server. (https://github.com/kubernetes/kubernetes/pull/112643)
    - Pod Security admission: the pod-security `warn` level will now default to the `enforce` level (https://github.com/kubernetes/kubernetes/pull/113491)
    - There is alpha feature gate `PodAndContainerStatsFromCRI` - returns container and pod metrics from CRI, instead of cAdvsior
    - Kube-proxy no longer falls back from ipvs mode to iptables mode if you ask it to do ipvs but the system is not correctly configured. Instead, it will just exit with an error. (https://github.com/kubernetes/kubernetes/pull/111806)
    - The `ServerSideApply` feature gate is locked to true with the feature already being GA. (https://github.com/kubernetes/kubernetes/pull/112748)
    - The `LegacyServiceAccountTokenNoAutoGeneration` feature gate is promoted to GA. But it is not yet locked to its default. The locking is planned for K8s 1.27. (https://github.com/kubernetes/kubernetes/pull/112838)
    
  - Release-independent tasks
    - Allow instantiating v1.26 Kubernetes clients and update `README.md` and `docs/usage/supported_k8s_versions.md`
    - Update feature gates in https://github.com/gardener/gardener/blob/master/pkg/utils/validation/features/featuregates.go
    - Update admission plugins in https://github.com/gardener/gardener/blob/master/pkg/utils/kubernetes/admissionplugins.go
    - Update KCM controllers in https://github.com/gardener/gardener/blob/master/pkg/operation/botanist/component/shootsystem/shootsystem.go
    - Bump used Kubernetes version for local shoot and local e2e test

  - Release-specific tasks
    - Enable kernel protection in `kubelet` ([ref](https://www.stigviewer.com/stig/kubernetes/2022-06-03/finding/V-242434), [ref](https://github.com/gardener/gardener/pull/6638#issuecomment-1266517345)): https://github.com/gardener/gardener/pull/6919
    - `kubelet` must not disable timeouts ([ref](https://www.stigviewer.com/stig/kubernetes/2022-06-03/finding/V-245541), [ref](https://github.com/gardener/gardener/pull/6638#issuecomment-1266517345)): https://github.com/gardener/gardener/pull/6937
    - Disable static kubeconfig for newly created `1.26+` shoot clusters
    - [CRI v1alpha2 removed](https://kubernetes.io/blog/2022/12/09/kubernetes-v1-26-release/#cri-v1alpha2-removed): https://github.com/gardener/gardener/pull/7265
    - CLI flag `pod-eviction-timeout` is deprecated and will be removed together with `enable-taint-manager` in v1.27. (https://github.com/kubernetes/kubernetes/pull/113710)
    - The metric `etcd_db_total_size_in_bytes` is renamed to `apiserver_storage_db_total_size_in_bytes`. (https://github.com/kubernetes/kubernetes/pull/113310)

## Test Infra

- [x] [Ensure test-infra compatibility](https://github.com/gardener/test-infra/blob/master/integration-tests/e2e/README.md#run-conformance-tests-against-new-k8s-versions) with conformance tests: https://github.com/gardener/test-infra/commit/231f14fd430cbfc76be8ae4f893b7fa0ddd484c5

## Cloud Controller Managers

- [x] `cloud-provider-aws`
  - Maintain `legacy-cloud-providers` and create `release-v1.26` branch from it (update container images and revendor latest Kubernetes patch): [`v1.26.1`](https://github.com/gardener/cloud-provider-gcp/releases/tag/v1.26.1)
  - Maintain `release-v1.25` branch (update container images and revendor latest Kubernetes patch): [`v1.25.5`](https://github.com/gardener/cloud-provider-aws/releases/tag/v1.25.5)
  - Maintain `release-v1.24` branch (update container images and revendor latest Kubernetes patch): [`v1.24.9`](https://github.com/gardener/cloud-provider-aws/releases/tag/v1.24.9)
  - Maintain `release-v1.23` branch (update container images and revendor latest Kubernetes patch): [`v1.23.15`](https://github.com/gardener/cloud-provider-aws/releases/tag/v1.23.15)
- [x] `cloud-provider-azure`
    - Maintain `release-v1.22` branch (update container images and revendor latest Kubernetes patch): [`v1.22.17`](https://github.com/gardener/cloud-provider-azure/releases/tag/v1.22.17)
  - Archive repository (no longer needed)
- [x] `cloud-provider-gcp`
  - Maintain `legacy-cloud-providers` and create `release-v1.26` branch from it (update container images and revendor latest Kubernetes patch): [`v1.26.1`](https://github.com/gardener/cloud-provider-gcp/releases/tag/v1.26.1)
  - Maintain `release-v1.25` branch (update container images and revendor latest Kubernetes patch): [`v1.25.5`](https://github.com/gardener/cloud-provider-gcp/releases/tag/v1.25.5)
  - Maintain `release-v1.24` branch (update container images and revendor latest Kubernetes patch): [`v1.24.9`](https://github.com/gardener/cloud-provider-gcp/releases/tag/v1.24.9)
  - Maintain `release-v1.23` branch (update container images and revendor latest Kubernetes patch): [`v1.23.15`](https://github.com/gardener/cloud-provider-gcp/releases/tag/v1.23.15)

## Provider Extensions

- [x] `provider-alicloud`: https://github.com/gardener/gardener-extension-provider-alicloud/pull/560
  - Revendor `gardener/gardener` with support for 1.26 client creation: https://github.com/gardener/gardener-extension-provider-alicloud/pull/565
  - Update `README.md` with support for 1.26
- [x] `provider-aws`: https://github.com/gardener/gardener-extension-provider-aws/pull/680
  - Revendor `gardener/gardener` with support for 1.26 client creation: https://github.com/gardener/gardener-extension-provider-aws/pull/695
  - Update `README.md` with support for 1.26
  - Update `cloud-controller-manager` images
- [x] `provider-azure`: https://github.com/gardener/gardener-extension-provider-azure/pull/619
  - Revendor `gardener/gardener` with support for 1.26 client creation: https://github.com/gardener/gardener-extension-provider-azure/pull/630
  - Update `README.md` with support for 1.26
  - Update `cloud-controller-manager` images
- [x] `provider-gcp`: https://github.com/gardener/gardener-extension-provider-gcp/pull/544
  - Revendor `gardener/gardener` with support for 1.26 client creation: https://github.com/gardener/gardener-extension-provider-gcp/pull/551
  - Update `README.md` with support for 1.26
  - Update `cloud-controller-manager` images
- [x] `provider-openstack`: https://github.com/gardener/gardener-extension-provider-openstack/pull/551
  - Revendor `gardener/gardener` with support for 1.26 client creation: https://github.com/gardener/gardener-extension-provider-openstack/pull/564
  - Update `README.md` with support for 1.26
  - Update `cloud-controller-manager` images
- [x] `provider-vsphere`: https://github.com/gardener/gardener-extension-provider-vsphere/pull/361
  - Revendor `gardener/gardener` with support for 1.26 client creation: https://github.com/gardener/gardener-extension-provider-vsphere/pull/373
  - Update `README.md` with support for 1.26
- [x] `provider-equinix-metal`: https://github.com/gardener/gardener-extension-provider-equinix-metal/pull/242
  - Revendor `gardener/gardener` with support for 1.26 client creation: https://github.com/gardener/gardener-extension-provider-equinix-metal/pull/244
  - Update `README.md` with support for 1.26

## Cluster Autoscaler

- [x] Update cluster-autoscaler for K8s 1.26: https://github.com/gardener/gardener/pull/7787

## Follow-up items

- [x] Remove the `enableStaticTokenKubeconfig` field after we figure out which tests/components need it and let the default gardener-apiserver value be used. Ref https://github.com/gardener/gardener/blob/f263358befa212f5c30c3fbc8e8e97479d2bf495/test/framework/resources/templates/default-shoot.yaml#L8-L13
  - https://github.com/gardener/gardener/pull/7495
  - https://github.com/gardener/gardener/pull/7536
  - https://github.com/gardener/gardener/pull/7551
  - https://github.com/gardener/gardener/pull/7609
- [x] Adapt the ManagedSeed controller to no longer rely on the static kubeconfig Secret: https://github.com/gardener/gardener/pull/7546

https://github.com/gardener/gardener/blob/e3dde16edf89fadf79c0e65a23273b6a9f759213/pkg/gardenlet/controller/managedseed/actuator.go#L511-L535

- [x] Adapt the e2e tests to do not enable the static token kubeconfig: https://github.com/gardener/gardener/pull/7617
- [x] Add K8s 1.26 support for Gardener to [kubernetes/test-infra](https://github.com/kubernetes/test-infra/): https://github.com/kubernetes/test-infra/pull/29024
- [x] Add K8s conformance badges to Gardener and extensions: https://github.com/gardener/gardener/pull/7655
- [x] Update the [the conformance spreadsheet](https://docs.google.com/spreadsheets/d/1uF9BoDzzisHSQemXHIKegMhuythuq_GL3N1mlUUK2h0/edit#gid=0&range=126:127) and reflect this update in our README.md: https://github.com/gardener/gardener/pull/7655

**Comment by yang-wang11** [comment]
@rfranzke Hello Rafael,
I noticed you mentioned the cloud-provider-azure will be archived, we still need its LB controller for our static IP.
May I know the reason? Is there any alternative controller responsible for reconciling the LB on the gardener further? Thanks.
```
 cloud-provider-azure
  Archive repository (no longer needed)
```

**Comment by kon-angelo** [comment]
@yang-wang11 for older releases we maintained a fork https://github.com/gardener/cloud-provider-azure of azure CCM but for newer k8s versions we use the upstream image.
TLDR; there will still be a CCM doing LB reconciliation (just not our fork).

**Comment by rfranzke** [comment]
We only use our k/k in-tree fork (https://github.com/gardener/cloud-provider-azure) for shoots <= 1.22. For shoot clusters higher than 1.22 Kubernetes, we already today use the upstream Azure cloud provider (github.com/kubernetes-sigs/cloud-provider-azure).

**Comment by rfranzke** [comment]
1.26 has been released last night.
/remove lifecycle/frozen

**Comment by himanshu-kun** [comment]
we should add `Sync g/autoscaler with upstream 1.26.0` as a task

**Comment by ialidzhikov** [comment]
/close
as all subtasks are now completed
